### PR TITLE
Use a new vars template implementation

### PR DESCRIPTION
**Pull request by lunny**
This PR reimplemented `Expand` in `modules/templates/vars` and replace all places.

### CODE REVIEW

**Review by techknowlogick — APPROVED** [review]

**Review by wxiaoguang — DISMISSED** [review]

File: modules/templates/vars/vars.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+			}
+			key.Reset()
+
+			enter = false
+		case enter:
+			if _, err := key.WriteRune(c); err != nil {
+				return "", err
+			}
+		default:
+			if _, err := buf.WriteRune(c); err != nil {
```
Comment (wxiaoguang) [inline modules/templates/vars/vars.go:99]: It could be `WriteByte` 

And I think we could support escaping in the new impelemtion, otherwise users couldn't write literal `{` and `}` chars. I would try to propose a PR.

File: models/migrations/migrations_test.go [hunk 2]
```diff
@@ -204,7 +203,7 @@ func prepareTestEnv(t *testing.T, skip int, syncModels ...interface{}) (*xorm.En
 	deferFn := PrintCurrentTest(t, ourSkip)
 	assert.NoError(t, os.RemoveAll(setting.RepoRootPath))
 
-	assert.NoError(t, com.CopyDir(path.Join(filepath.Dir(setting.AppPath), "integrations/gitea-repositories-meta"),
+	assert.NoError(t, util.CopyDir(path.Join(filepath.Dir(setting.AppPath), "integrations/gitea-repositories-meta"),
```
Comment (wxiaoguang) [inline models/migrations/migrations_test.go]: We can see that there was a PR:
* #14103

@6543 has tried to refactor the `com.Xxx` before, but these legacy functions still appears in new code ..... should we write the `com` into block list in lint rules?

Comment (lunny) [reply models/migrations/migrations_test.go]: see https://gitea.com/gitea/gitea-vet/pulls/20

File: modules/templates/vars/vars.go (hunk 1 above)
Comment (lunny) [reply modules/templates/vars/vars.go:99]: I think we have to use `WriteRune` because `c` is a `rune` type but not `byte` here.
I just implement it minimal to match our current requirement. We can extend it when we have further requirements.

Comment (wxiaoguang) [reply modules/templates/vars/vars.go:99]: Oh, sorry, I misread the code. c is a Rune.

Could you wait for a few minutes? I think we could make the implement better.

And the code `v = subs[0]` seem incorrect either.

Comment (wxiaoguang) [reply modules/templates/vars/vars.go:99]: There is one more thing, in some cases, we should just return the partially rendered template content with the error, and the caller can ignore the error just show the content to end users. we should not always show 500 to end users.

File: models/migrations/migrations_test.go (hunk 2 above)
Comment (wxiaoguang) [reply models/migrations/migrations_test.go]: I think we do not need it.

image


My PR is almost done.

File: modules/templates/vars/vars.go (hunk 1 above)
Comment (lunny) [reply modules/templates/vars/vars.go:99]: I think we need to return the error clearly. I cannot find the use case to need partially rendered template.

**Comment by wxiaoguang** [comment]
PR: 
* https://github.com/lunny/gitea/pull/22

Changes:
1. Add `com` to deny list
2. Improve `vars.Expand`
    1. now it always returns the rendered string (even if there is an error)
    2. now the key is allocation-free, and bytes are written in batches, faster.
    3. support escaped chars, and now we have the ability to extend the syntax in future by the reserved char `#`
4. In some cases, always show rendered text to end users, avoid unnecessary 500 erros
5. Fix some errors in golangci.yml

Maybe the title of this PR could be changed accordingly: `Refactor legacy com(common) package and improve the simple string template Expand`

**Comment by lunny** [comment]
> PR:
> 
> * [Remove most calls to `com`, improve `vars.Expand` lunny/gitea#22](https://github.com/lunny/gitea/pull/22)
> 
> Changes:
> 
> 1. Add `com` to deny list
This could be another PR.
> 2. Improve `vars.Expand`
>    
>    1. it always returns the rendered string (even if there is an error)
I prefer a strict mode.
>    2. the key processing is allocation-free, and bytes are written in string batches, faster.
I have changed my code, now it's also allocation-free for key processing.
>    3. support escaped chars, and now we have the ability to extend the syntax in future by the reserved char `#`
Why we need that? The key should not contain `{`.
> 3. In some cases, always show rendered text to end users, avoid unnecessary 500 erros
Hiding the error means we catch the error more later.
> 4. Fix some errors in golangci.yml
This could be another PR.
> 
> Maybe the title of this PR could be changed accordingly: `Refactor legacy com(common) package and improve the simple string template Expand`
I will remove all changes about `com` to keep this PR clean.

File: modules/context/repo.go [hunk 3]
```diff
@@ -309,11 +309,17 @@ func EarlyResponseForGoGetMeta(ctx *Context) {
...
 			"GoGetImport": ComposeGoGetImport(username, reponame),
 			"CloneLink":   repo_model.ComposeHTTPSCloneURL(username, reponame),
-		}))
+		})
+	if err != nil {
+		log.Error(err.Error())
+		ctx.PlainText(http.StatusInternalServerError, "expand vars failed")
+		return
+	}
+	ctx.PlainText(http.StatusOK, res)
```
Comment (delvh) [inline modules/context/repo.go]: I just noticed: That whole operation is overkill. It can be greatly simplified:
```suggestion
	ctx.PlainText(http.StatusOK, fmt.Sprintf(`<meta name="go-import" content="%s git %s">`, ComposeGoGetImport(username, reponame), repo_model.ComposeHTTPSCloneURL(username, reponame)))
```

File: modules/context/repo.go [hunk 4]
```diff
@@ -26,11 +26,11 @@ import (
 	"code.gitea.io/gitea/modules/markup/markdown"
 	"code.gitea.io/gitea/modules/setting"
 	api "code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/modules/templates/vars"
```
Comment (delvh) [inline modules/context/repo.go]:
```suggestion
```

File: routers/web/repo/issue.go [hunk 5]
```diff
@@ -1103,7 +1102,13 @@ func ViewIssue(ctx *context.Context) {
 			if extIssueUnit.ExternalTrackerConfig().ExternalTrackerStyle == markup.IssueNameStyleNumeric || extIssueUnit.ExternalTrackerConfig().ExternalTrackerStyle == "" {
 				metas := ctx.Repo.Repository.ComposeMetas()
 				metas["index"] = ctx.Params(":index")
-				ctx.Redirect(com.Expand(extIssueUnit.ExternalTrackerConfig().ExternalTrackerFormat, metas))
+				res, err := vars.Expand(extIssueUnit.ExternalTrackerConfig().ExternalTrackerFormat, metas)
+				if err != nil {
+					log.Error("unable to expand template vars for issue url. issue: %s, err: %s", metas["index"], err.Error())
+					ctx.ServerError("Expand", err)
```
Comment (delvh) [inline routers/web/repo/issue.go]: Isn't that somewhat duplicated? Because `ctx.ServerError` already displays the error in the logs, but without the actual cause.

File: routers/web/goget.go [hunk 6]
```diff
@@ -67,7 +67,7 @@ func goGet(ctx *context.Context) {
 	}
 	ctx.RespHeader().Set("Content-Type", "text/html")
 	ctx.Status(http.StatusOK)
-	_, _ = ctx.Write([]byte(com.Expand(`<!doctype html>
+	res, err := vars.Expand(`<!doctype html>
```
Comment (delvh) [inline routers/web/goget.go]:
```suggestion
	goGetImport := context.ComposeGoGetImport(ownerName, trimmedRepoName)
	goDocDirectory := prefix + "{/dir}"
	ctx.Write([]byte(fmt.Sprintf(`<!doctype html>
<html>
	<head>
		<meta name="go-import" content="%s git %s">
		<meta name="go-source" content="%s _ %s %s">
	</head>
	<body>
		go get %s
	</body>
</html>
`, goGetImport, repo_model.ComposeHTTPSCloneURL(ownerName, repoName), goGetImport, goDocDirectory, goDocDirectory + "/{file}#L{line}", insecure+goGetImport)))
```
(but it's probably better if you copy this suggestion into your IDE and manually format it, as it's not unlikely that I made a formatting error, and also I can't throw out the lines below for some reason...)

File: modules/templates/vars/vars.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+package vars
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ErrWrongSyntax represents a wrong syntax with a tempate
+type ErrWrongSyntax struct {
+	Template string
```
Comment (delvh) [inline modules/templates/vars/vars.go:14]: I think it might be beneficial to serve a reason as well in this case, i.e. `Reason: "\"{\" is not allowed inside another match"`:
```suggestion
	Template string
	Reason string
```

File: modules/templates/vars/vars.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func (err ErrWrongSyntax) Error() string {
+	return fmt.Sprintf("Wrong syntax found in %s", err.Template)
```
Comment (delvh) [inline modules/templates/vars/vars.go]:
```suggestion
	return fmt.Sprintf("Wrong syntax found in %s: %s", err.Template, err.Reason)
```

File: modules/templates/vars/vars.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+// license that can be found in the LICENSE file.
+
+package vars
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ErrWrongSyntax represents a wrong syntax with a tempate
```
Comment (delvh) [inline modules/templates/vars/vars.go]:
```suggestion
// ErrWrongSyntax represents a syntax error within a template
```

File: modules/templates/vars/vars.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// IsErrWrongSyntax returns true if the error is ErrWrongSyntax
+func IsErrWrongSyntax(err error) bool {
+	_, ok := err.(ErrWrongSyntax)
+	return ok
+}
+
+// ErrNoMatchedVar represents an error that no matched vars
+type ErrNoMatchedVar struct {
```
Comment (delvh) [inline modules/templates/vars/vars.go:29]: Is that really an error?
How do you pass anything inside curly braces then that is not supposed to be a variable then?

File: modules/templates/vars/vars.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	var (
+		buf         strings.Builder
+		keyStartPos = -1
+	)
+	for i, c := range template {
+		switch {
+		case c == '{':
+			if keyStartPos > -1 {
+				return "", ErrWrongSyntax{
+					Template: template,
```
Comment (delvh) [inline modules/templates/vars/vars.go:55]:
```suggestion
					Template: template,
					Reason: "\"{\" is not allowed to occur again before closing the variable",
```

File: modules/templates/vars/vars.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+				}
+			}
+			keyStartPos = i
+		case c == '}':
+			if keyStartPos == -1 {
+				return "", ErrWrongSyntax{
+					Template: template,
```
Comment (delvh) [inline modules/templates/vars/vars.go:63]:
```suggestion
					Template: template,
					Reason: "\"}\" can only occur after an opening \"{\"",
```

File: modules/templates/vars/vars.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+				}
+			}
+			if i-keyStartPos <= 1 {
+				return "", ErrWrongSyntax{
+					Template: template,
+				}
+			}
+
+			if len(match) == 0 {
```
Comment (delvh) [inline modules/templates/vars/vars.go:74]: 1. Why is that an error case? I strongly disagree that not finding the corresponding variable should return an error.
2. Even if I count that as an error case, it is an unnecessary one as that will already be tested down below with the `v, ok := match[template[keyStartPos+1:i]]` implicitly.
3. Even more so, it is implemented incorrectly as it disregards the positional arguments.

File: modules/templates/vars/vars.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+			keyStartPos = i
+		case c == '}':
+			if keyStartPos == -1 {
+				return "", ErrWrongSyntax{
+					Template: template,
+				}
+			}
+			if i-keyStartPos <= 1 {
+				return "", ErrWrongSyntax{
+					Template: template,
```
Comment (delvh) [inline modules/templates/vars/vars.go:69]:
```suggestion
					Template: template,
					Reason: "the empty variable (\"{}\") is not allowed",
```

File: modules/templates/vars/vars.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+			}
+
+			keyStartPos = -1
+		case keyStartPos > -1:
+		default:
+			if _, err := buf.WriteRune(c); err != nil {
+				return "", err
+			}
+		}
+	}
```
Comment (delvh) [inline modules/templates/vars/vars.go:103]: Currently, there is an edge case missing, as an unclosed parentheses at the end (i.e. `test {template end` as template) would not throw a syntax error.

File: modules/templates/vars/vars_test.go [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+		{
+			template: "expand {{a}, {b} and {c}",
+			fail:     true,
+		},
+		{
+			template: "expand {}, {b} and {c}",
+			fail:     true,
+		},
+		{
+			template: "expand }, {b} and {c}",
```
Comment (delvh) [inline modules/templates/vars/vars_test.go:54]: As mentioned above, the other case is currently missing.

File: modules/templates/vars/vars.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// IsErrNoMatchedVar returns true if the error is ErrNoMatchedVar
+func IsErrNoMatchedVar(err error) bool {
+	_, ok := err.(ErrNoMatchedVar)
+	return ok
+}
+
+// Expand replaces all variables like {var} to match
+func Expand(template string, match map[string]string, subs ...string) (string, error) {
```
Comment (delvh) [inline modules/templates/vars/vars.go:45]: You could save all the extra trouble with the positional arguments by using something like
```suggestion
// All positional arguments will be converted into corresponding map entries: match["x"] = subs[x])
func Expand(template string, match map[string]string, subs ...string) (string, error) {
  for index, substitution := range subs {
    match[strconv.Itoa(index)] = substitution
  }
```

File: routers/web/goget.go (hunk 6 above)
Comment (lunny) [reply routers/web/goget.go]: I use `os.Expand` to replace `vars.Expand` here.

**Comment by lunny** [comment]
replaced by #19284
